File: src/bin/0416-partition-equal-subset-sum.py
# @leetup=info id=416 lang=python3 slug=partition-equal-subset-sum



# @leetup=code
class Solution:
    def canPartition(self, nums: list[int]) -> bool:
        N = sum(nums) + 1
        dp = [False] * N
        dp[0] = True

        for n in nums:
            next_dp = [False] * N
            for m, possible in enumerate(dp):
                if not possible:
                    continue
                next_dp[abs(m + n)] = True
                next_dp[abs(m - n)] = True
            dp = next_dp

        return dp[0]


# @leetup=code

File: src/bin/0416-partition-equal-subset-sum.rs
// @leetup=info id=416 lang=rust slug=partition-equal-subset-sum

// @leetup=code
impl Solution {
    pub fn can_partition(nums: Vec<i32>) -> bool {
        let len = (nums.iter().cloned().sum::<i32>() + 1) as usize;
        let mut dp = vec![false; len];
        dp[0] = true;

        for n in nums {
            let n = n as usize;
            let mut next_dp = vec![false; len];
            // for m in dp.into_iter().enumerate().filter(|(_, p)| *p).map(|(i, _)| i) {
            for m in dp
                .into_iter()
                .enumerate()
                .flat_map(|(m, p)| if p { Some(m) } else { None })
            {
                next_dp[(m + n)] = true;

                // Unstable:
                // next_dp[m.abs_diff(n)] = true;

                if m > n {
                    next_dp[m - n] = true;
                } else {
                    next_dp[n - m] = true;
                }
            }
            dp = next_dp;
        }

        dp[0]
    }
}
// @leetup=code

struct Solution;

fn main() {
    let _ = Solution::can_partition;
}
